File: solutions/cpp/0332.cpp
class Solution {
 public:
  vector<string> findItinerary(vector<vector<string>>& tickets) {
    deque<string> ans;
    map<string, multiset<string>> graph;

    for (const vector<string>& ticket : tickets)
      graph[ticket[0]].insert(ticket[1]);

    function<void(const string&)> dfs = [&](const string& airport) {
      while (graph.count(airport) && !graph[airport].empty()) {
        const string nextAirport = *begin(graph[airport]);
        graph[airport].erase(begin(graph[airport]));
        dfs(nextAirport);
      }
      ans.push_front(airport);
    };

    dfs("JFK");

    return {begin(ans), end(ans)};
  }
};

File: solutions/java/0332.java
class Solution {
  public List<String> findItinerary(List<List<String>> tickets) {
    List<String> ans = new LinkedList<>();
    Map<String, PriorityQueue<String>> graph = new HashMap<>();

    for (final List<String> ticket : tickets)
      graph.computeIfAbsent(ticket.get(0), k -> new PriorityQueue<>()).add(ticket.get(1));

    dfs("JFK", graph, ans);

    return ans;
  }

  private void dfs(final String airport, Map<String, PriorityQueue<String>> graph, List<String> ans) {
    while (graph.containsKey(airport) && !graph.get(airport).isEmpty()) {
      String nextAirport = graph.get(airport).poll();
      dfs(nextAirport, graph, ans);
    }
    ans.add(0, airport);
  }
}